### ISSUE TITLE
error in runing both "mb" and "glasso" method: invalid class “dgCMatrix” object: all row indices must be between 0 and nrow-1

### ISSUE THREAD

**Issue by Lei0509** [zdk123/SpiecEasi#111]
Hi Zachary, 

This is Lei
I have an issue to running the spiec.easi function. I think it should be something related to my sampledata. When I apply this function to my whole dataset (40 samples), it works. However, when select the subsample (5 samples), it doesn't work.. I try to modify the parameters of the function but it still did not work. 

Please find the error message below:
>   spiec <- spiec.easi(Bact.occ8, method='mb', nlambda=100,
+                       lambda.min.ratio=1e-2, pulsar.params = list(rep.num=50))
Applying data transformations...
Selecting model with pulsar using stars...
### Error in validObject(.Object) : 
  invalid class “dgCMatrix” object: all row indices must be between 0 and nrow-1

Can you please help me to solve this issues. Thank you so much!!

**Comment by zdk123** [comment]
Could be a pathological case when sample size if very small. How many OTUs do you have? I would need to see the data or a another case that reproduces the error. 

I agree this error message is not very helpful, but I haven't seen it before.

**Comment by Lei0509** [comment]
Hi @zdk123 , 
Thank you so much for your response. 
For my entire dataset I have 1100 OTU, but for the subsample, I only have about 300. 
[A2017fall_tax.txt](https://github.com/zdk123/SpiecEasi/files/3748620/A2017fall_tax.txt)
[16s_otu_silva_norare_phyloseq_otu.txt](https://github.com/zdk123/SpiecEasi/files/3748622/16s_otu_silva_norare_phyloseq_otu.txt)

I attached my files in above. Can you please have a look. Thank you. !

**Comment by zdk123** [comment]
Thanks!  am able to reproduce the error via the following:
```
X <- t(read.table("16s_otu_silva_norare_phyloseq_otu.txt", sep="\t", header=TRUE, row.names=1))
set.seed(10010)
ind <- sample(1:nrow(X), 5)
cind <- colSums(sign(X[ind,]))>=3

library(SpiecEasi)
est <- spiec.easi(X[ind,cind], method='mb')
```
I'll look into this.

**Comment by zdk123** [comment]
I can avoid the bug by bumping up `lambda.min.ratio` to something 1e-1... can you try that? Looks like either an empty path and /or segfault from huge.mb.

**Comment by Lei0509** [comment]
Hi @zdk123 , 
Thank you for your help. 
1) I have tried increase lambda.min.ratio before. Depending on the subsample I have, it works sometimes. What is the maximum lambda.min.ratio I could have??

2) Another problem is that after increasing the  lambda.min.ratio, I got the result but there are warning message saying that "**na introduced by coercion"**. If I used this result for plotting by igraph, that won't work. The error message is like "**the name of vertex does not exit (or invalid)**".  (I cannot remember the exact message. There is not graph produced.)

3) Increased the lambda.min.ratio did not work for one of specific subsample which I am working on now. Can you please try the subsample using the file which I attached that included the name showed below: 
**GO1_1
GO1_2
GO2_2
GO3_1
GO4_1
GO4_2**

4) Another questions is that should I used same parameter setting for all the subsample to keep consistent?

**Comment by zdk123** [comment]
It's definitely a sample size issue, but I don't have a good heuristic for you.  The lambda penalty interacts with the sample size via the sample correlation matrix. 
So you'd need to control for the the noise in the correlation matrix, the sample size, number of OTUs and the magnitude of the true correlations, plus other sources of statistical error.

What you may be seeing is stability issues in the MB method when the dataset is extremely noisy. Especially since to do model selection, we run 80% subsamples through the method... so 5 full samples is really only 4 samples. And since 5 choose 4 = 5, you can't actually get a useful StARS summary statistic anyway.

 I would only trying network inference over hundreds of samples, and probably no fewer than 40.

**Comment by Lei0509** [comment]
Hi @zdk123 , 

Thank you so much for your explaination. 
I have one more question. **What is the maximum and minimum lambda.min.ratio** I can use for the "MB" method. **Will different lambda.min.ratio can significantly affect the result**? (Or how it can affect the result?)

Thank you so much for your help.

**Comment by zdk123** [comment]
Yeah I really need to include this in the troubleshooting / FAQ section. As a mental approximation, the value of lambda is related to the sparsity of the network (though the relationship is not linear). The very large lambdas will result in an empty graph while very small lambdas will result in a completely connected graph.

For the Gaussian model, the maximum lambda is the absolute value of the largest [in magnitude] correlation of the sample correlation matrix. Actually you _can_ supply larger lambdas, but the graph can't be any emptier than zero edges. Similarly, lambdas can be arbitrarily small (as long as they are non-zero). but there's no point in trying extremely small lambdas if you have already found the full graph.

There are a few other considerations as well: computing graphical models with small values of lambda is slow (see our [preprint on this](https://arxiv.org/abs/1605.07072)) and, since you usually need regularization to compute an invertible precision matrix, the results may not be stable.

What we do in StARS is make a lambda path where we have parameters `lambda.max = |correlation.max|`, `lambda.min = lambda.min.ratio * lambda.max` and `nlambda`, where we have a vector of `nlambda` lambdas equally spaced between `lambda.min` and `lambda.max`. Then, we solve for a Gaussian graphical model at each value of lambda, and find the optimal lambda via the StARS procedure. 

So in principle, the graph at lambda_i could be independent of the graph at lambda_i+1. So if we try smaller or larger of values of lambda, the optimal graph will be the same as long as it is in our lambda path at all. However, if the values are too small we get speed and stability problems as you are seeing.

In practice, however, the graphs solved at each lambda are not independent... to get a speed boost, we can typically use solution at lambda_i to "seed" the initialization of lambda_i+1 (warm starts). Since the objective function is convex, this should converge to the same solution, but in practice this doesn't always always happen (e.g. depending on the stopping criteria, step size, etc, in the underlying algorithm).

What I typically recommend is leaving the lambda.max alone (since this is fixed in theory, and large values of lambda are computationally cheap) and starting with a large lambda.min.ratio (something like 1e-1, a small nlambda and a small rep.num. Incrementally decrease the former and increase the latter two parameters until the StARS summary statistic is as close as possible to the threshold. Increasing `rep.num` increases the number of StARS subsamples, which helps with robustness to outliers and precision of the StARS statistic.

**Comment by Lei0509** [comment]
Hi @zdk123 , 

Thank you so much for your detail explanation and your patience. It really helps me to have better understanding of the "mb" method in the SpiecEasi. 

Yes, it will be great if you can include the explanation here to the trouble shooting or the  FAQ section in the main SpiecEasi page so that everyone can this information and start their test. 

Thank you!
